File: sdk/java-holons/src/test/java/org/organicprogramming/holons/DiscoveryTestSupport.java
package org.organicprogramming.holons;

import com.google.gson.Gson;
import org.organicprogramming.holons.DiscoveryTypes.DiscoverResult;
import org.organicprogramming.holons.DiscoveryTypes.HolonInfo;
import org.organicprogramming.holons.DiscoveryTypes.HolonRef;
import org.organicprogramming.holons.DiscoveryTypes.IdentityInfo;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

final class DiscoveryTestSupport {
    private static final Gson GSON = new Gson();

    private DiscoveryTestSupport() {
    }

    static RuntimeFixture runtimeFixture(Path root) throws IOException {
        Path opHome = root.resolve("runtime");
        Path opBin = opHome.resolve("bin");
        Files.createDirectories(opBin);
        return new RuntimeFixture(root, opHome, opBin);
    }

    static RuntimeEnv runtimeEnv(RuntimeFixture fixture) {
        return new RuntimeEnv(fixture.root(), fixture.opHome(), fixture.opBin());
    }

    static List<String> slugs(DiscoverResult result) {
        List<String> slugs = new ArrayList<>();
        if (result == null || result.found == null) {
            return slugs;
        }
        for (HolonRef ref : result.found) {
            if (ref != null && ref.info != null) {
                slugs.add(ref.info.slug);
            }
        }
        slugs.sort(String::compareTo);
        return slugs;
    }

    static void writePackageHolon(Path dir, PackageSeed seed) throws IOException {
        Files.createDirectories(dir);

        String slug = seed.slug != null && !seed.slug.isBlank()
                ? seed.slug
                : slugFor(seed.givenName, seed.familyName);
        String runner = seed.runner != null && !seed.runner.isBlank() ? seed.runner : "go-module";
        String kind = seed.kind != null && !seed.kind.isBlank() ? seed.kind : "native";

        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("schema", "holon-package/v1");
        payload.put("slug", slug);
        payload.put("uuid", seed.uuid);

        Map<String, Object> identity = new LinkedHashMap<>();
        identity.put("given_name", seed.givenName);
        identity.put("family_name", seed.familyName);
        identity.put("motto", seed.motto == null ? "" : seed.motto);
        identity.put("aliases", seed.aliases == null ? List.of() : seed.aliases);
        payload.put("identity", identity);

        payload.put("lang", seed.lang == null ? "go" : seed.lang);
        payload.put("runner", runner);
        payload.put("status", seed.status == null ? "draft" : seed.status);
        payload.put("kind", kind);
        payload.put("transport", seed.transport == null ? "" : seed.transport);
        payload.put("entrypoint", seed.entrypoint == null ? slug : seed.entrypoint);
        payload.put("architectures", seed.architectures == null ? List.of() : seed.architectures);
        payload.put("has_dist", seed.hasDist);
        payload.put("has_source", seed.hasSource);

        Files.writeString(dir.resolve(".holon.json"), GSON.toJson(payload) + System.lineSeparator());
    }

    static HolonInfo holonInfo(String slug, String uuid, String givenName, String familyName) {
        HolonInfo info = new HolonInfo();
        info.slug = slug;
        info.uuid = uuid;
        info.identity = new IdentityInfo();
        info.identity.givenName = givenName;
        info.identity.familyName = familyName;
        info.identity.aliases = new ArrayList<>();
        info.lang = "go";
        info.runner = "go-module";
        info.status = "draft";
        info.kind = "native";
        return info;
    }

    static HolonRef holonRef(Path path, String slug, String uuid, String givenName, String familyName) {
        HolonRef ref = new HolonRef();
        ref.url = path.toAbsolutePath().normalize().toUri().toString();
        ref.info = holonInfo(slug, uuid, givenName, familyName);
        return ref;
    }

    static String platformTag() {
        return normalizedOs() + "_" + normalizedArch();
    }

    static String slugFor(String givenName, String familyName) {
        return (safe(givenName) + "-" + safe(familyName))
                .trim()
                .toLowerCase()
                .replace(" ", "-")
                .replaceAll("^-+|-+$", "");
    }

    static void writeExecutable(Path path, String content) throws IOException {
        Files.createDirectories(path.getParent());
        Files.writeString(path, content);
        if (!path.toFile().setExecutable(true)) {
            throw new IOException("failed to mark executable: " + path);
        }
    }

    private static String normalizedOs() {
        String os = System.getProperty("os.name", "").toLowerCase();
        if (os.contains("mac") || os.contains("darwin")) {
            return "darwin";
        }
        if (os.contains("win")) {
            return "windows";
        }
        return "linux";
    }

    private static String normalizedArch() {
        String arch = System.getProperty("os.arch", "").toLowerCase();
        if ("aarch64".equals(arch) || "arm64".equals(arch)) {
            return "arm64";
        }
        if ("x86_64".equals(arch) || "amd64".equals(arch)) {
            return "amd64";
        }
        return arch.replace('-', '_');
    }

    private static String safe(String value) {
        return value == null ? "" : value;
    }

    record RuntimeFixture(Path root, Path opHome, Path opBin) {
    }

    static final class RuntimeEnv implements AutoCloseable {
        private final String previousUserDir = System.getProperty("user.dir");
        private final String previousOpPath = System.getProperty("OPPATH");
        private final String previousOpBin = System.getProperty("OPBIN");

        RuntimeEnv(Path root, Path opHome, Path opBin) {
            System.setProperty("user.dir", root.toString());
            System.setProperty("OPPATH", opHome.toString());
            System.setProperty("OPBIN", opBin.toString());
        }

        @Override
        public void close() {
            restore("user.dir", previousUserDir);
            restore("OPPATH", previousOpPath);
            restore("OPBIN", previousOpBin);
        }

        private static void restore(String key, String value) {
            if (value == null) {
                System.clearProperty(key);
            } else {
                System.setProperty(key, value);
            }
        }
    }

    static final class PackageSeed {
        String slug;
        String uuid;
        String givenName;
        String familyName;
        String motto = "";
        String lang = "go";
        String runner = "go-module";
        String status = "draft";
        String kind = "native";
        String transport = "";
        String entrypoint;
        List<String> architectures = List.of();
        boolean hasDist;
        boolean hasSource;
        List<String> aliases = List.of();

        PackageSeed(String uuid, String givenName, String familyName) {
            this.uuid = uuid;
            this.givenName = givenName;
            this.familyName = familyName;
        }

        PackageSeed slug(String value) {
            this.slug = value;
            return this;
        }

        PackageSeed entrypoint(String value) {
            this.entrypoint = value;
            return this;
        }

        PackageSeed aliases(List<String> value) {
            this.aliases = value;
            return this;
        }

        PackageSeed transport(String value) {
            this.transport = value;
            return this;
        }

        PackageSeed hasDist(boolean value) {
            this.hasDist = value;
            return this;
        }

        PackageSeed hasSource(boolean value) {
            this.hasSource = value;
            return this;
        }
    }
}

File: sdk/kotlin-holons/src/test/kotlin/org/organicprogramming/holons/DiscoveryTestSupport.kt
package org.organicprogramming.holons

import io.grpc.CallOptions
import io.grpc.ManagedChannel
import io.grpc.MethodDescriptor
import io.grpc.stub.ClientCalls
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import java.io.ByteArrayInputStream
import java.io.IOException
import java.io.InputStream
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.time.Duration
import java.util.concurrent.TimeUnit
import kotlin.test.assertTrue

private val sdkRoot: Path = Path.of("").toAbsolutePath().normalize()
private val json = Json
private val pingMethod: MethodDescriptor<String, String> = MethodDescriptor.newBuilder<String, String>()
    .setType(MethodDescriptor.MethodType.UNARY)
    .setFullMethodName(MethodDescriptor.generateFullMethodName("echo.v1.Echo", "Ping"))
    .setRequestMarshaller(stringMarshaller())
    .setResponseMarshaller(stringMarshaller())
    .build()

internal data class RuntimeFixture(
    val root: Path,
    val opHome: Path,
    val opBin: Path,
)

internal data class PackageSeed(
    val slug: String,
    val uuid: String,
    val givenName: String,
    val familyName: String,
    val runner: String = "go-module",
    val entrypoint: String,
    val kind: String = "native",
    val aliases: List<String> = emptyList(),
    val architectures: List<String> = emptyList(),
    val hasDist: Boolean = false,
    val hasSource: Boolean = false,
    val transport: String = "",
)

internal data class ConnectFixture(
    val slug: String,
    val sourceDir: Path,
    val pidFile: Path,
    val argsFile: Path,
    val info: HolonInfo,
)

internal fun <T> withRuntimeFixture(block: (RuntimeFixture) -> T): T {
    val root = Files.createTempDirectory("kotlin-holons-runtime-")
    val opHome = root.resolve("runtime")
    val opBin = opHome.resolve("bin")

    val previousUserDir = System.getProperty("user.dir")
    val previousOpPath = System.getProperty("OPPATH")
    val previousOpBin = System.getProperty("OPBIN")
    val previousSiblingsRoot = System.getProperty("holons.siblings.root")
    val previousSourceOffload = sourceDiscoverOffload
    val previousProbeOverride = packageDescribeProbeOverride

    System.setProperty("user.dir", root.toString())
    System.setProperty("OPPATH", opHome.toString())
    System.setProperty("OPBIN", opBin.toString())
    System.clearProperty("holons.siblings.root")

    sourceDiscoverOffload = { _, _, _, _, _, _ -> DiscoverResult() }
    packageDescribeProbeOverride = null

    return try {
        block(RuntimeFixture(root, opHome, opBin))
    } finally {
        restoreProperty("user.dir", previousUserDir)
        restoreProperty("OPPATH", previousOpPath)
        restoreProperty("OPBIN", previousOpBin)
        restoreProperty("holons.siblings.root", previousSiblingsRoot)
        sourceDiscoverOffload = previousSourceOffload
        packageDescribeProbeOverride = previousProbeOverride
        root.toFile().deleteRecursively()
    }
}

internal fun restoreProperty(name: String, value: String?) {
    if (value == null) {
        System.clearProperty(name)
    } else {
        System.setProperty(name, value)
    }
}

internal fun sortedSlugs(result: DiscoverResult): List<String> =
    result.found.mapNotNull { it.info?.slug }.sorted()

internal fun writePackageHolon(dir: Path, seed: PackageSeed) {
    Files.createDirectories(dir)
    val aliases = if (seed.aliases.isEmpty()) "[]" else seed.aliases.joinToString(prefix = "[", postfix = "]") { "\"$it\"" }
    val architectures = if (seed.architectures.isEmpty()) "[]" else seed.architectures.joinToString(prefix = "[", postfix = "]") { "\"$it\"" }
    Files.writeString(
        dir.resolve(".holon.json"),
        """
        {
          "schema": "holon-package/v1",
          "slug": "${seed.slug}",
          "uuid": "${seed.uuid}",
          "identity": {
            "given_name": "${seed.givenName}",
            "family_name": "${seed.familyName}",
            "aliases": $aliases
          },
          "lang": "kotlin",
          "runner": "${seed.runner}",
          "status": "draft",
          "kind": "${seed.kind}",
          "transport": "${seed.transport}",
          "entrypoint": "${seed.entrypoint}",
          "architectures": $architectures,
          "has_dist": ${seed.hasDist},
          "has_source": ${seed.hasSource}
        }
        """.trimIndent() + "\n",
    )
}

internal fun writeStaticDescribePackage(dir: Path) {
    val binaryDir = dir.resolve("bin").resolve(currentPlatformDirName())
    Files.createDirectories(binaryDir)
    val wrapper = binaryDir.resolve("static-wrapper")
    Files.writeString(
        wrapper,
        """
        #!/usr/bin/env bash
        set -euo pipefail
        if [[ "${'$'}{1:-}" == "serve" ]]; then
          shift
        fi
        if [[ "${'$'}{1:-}" == "--listen" ]]; then
          shift
        fi
        exec ${shellQuote(Path.of(System.getProperty("java.home"), "bin", "java").toString())} -cp ${shellQuote(System.getProperty("java.class.path"))} org.organicprogramming.holons.StaticDescribeServerMain "${'$'}{1:-stdio://}"
        """.trimIndent() + "\n",
    )
    assertTrue(wrapper.toFile().setExecutable(true), "static wrapper should be executable")
}

internal fun createConnectFixture(root: Path, givenName: String, familyName: String): ConnectFixture {
    val slug = "$givenName-$familyName".lowercase()
    val sourceDir = root.resolve("holons").resolve(slug)
    val binaryDir = sourceDir.resolve(".op").resolve("build").resolve("bin")
    Files.createDirectories(binaryDir)

    val wrapper = binaryDir.resolve("echo-wrapper")
    val pidFile = root.resolve("$slug.pid")
    val argsFile = root.resolve("$slug.args")

    Files.writeString(
        wrapper,
        """
        #!/usr/bin/env bash
        set -euo pipefail
        printf '%s\n' "${'$'}${'$'}" > ${shellQuote(pidFile.toString())}
        : > ${shellQuote(argsFile.toString())}
        for arg in "${'$'}@"; do
          printf '%s\n' "${'$'}arg" >> ${shellQuote(argsFile.toString())}
        done
        exec ${shellQuote(sdkRoot.resolve("bin").resolve("echo-server").toString())} "${'$'}@"
        """.trimIndent() + "\n",
    )
    assertTrue(wrapper.toFile().setExecutable(true), "echo wrapper should be executable")

    val info = HolonInfo(
        slug = slug,
        uuid = "$slug-uuid",
        identity = IdentityInfo(givenName = givenName, familyName = familyName),
        lang = "kotlin",
        runner = "shell",
        status = "draft",
        kind = "service",
        transport = "",
        entrypoint = "echo-wrapper",
        architectures = emptyList(),
        hasDist = true,
        hasSource = true,
    )

    return ConnectFixture(slug = slug, sourceDir = sourceDir, pidFile = pidFile, argsFile = argsFile, info = info)
}

internal fun connectFixtureRef(fixture: ConnectFixture): HolonRef =
    HolonRef(url = fileURL(fixture.sourceDir), info = fixture.info)

internal fun waitForPidFile(path: Path): Long {
    val deadline = System.nanoTime() + Duration.ofSeconds(5).toNanos()
    while (System.nanoTime() < deadline) {
        try {
            val pid = Files.readString(path).trim().toLong()
            if (pid > 0) {
                return pid
            }
        } catch (_: Exception) {
            // Still starting.
        }
        Thread.sleep(25)
    }
    error("timed out waiting for pid file $path")
}

internal fun waitForArgsFile(path: Path): List<String> {
    val deadline = System.nanoTime() + Duration.ofSeconds(5).toNanos()
    while (System.nanoTime() < deadline) {
        try {
            val lines = Files.readAllLines(path, StandardCharsets.UTF_8).filter { it.isNotBlank() }
            if (lines.isNotEmpty()) {
                return lines
            }
        } catch (_: IOException) {
            // Still starting.
        }
        Thread.sleep(25)
    }
    error("timed out waiting for args file $path")
}

internal fun waitForPidExit(pid: Long) {
    val deadline = System.nanoTime() + Duration.ofSeconds(5).toNanos()
    while (System.nanoTime() < deadline) {
        val alive = ProcessHandle.of(pid).map(ProcessHandle::isAlive).orElse(false)
        if (!alive) {
            return
        }
        Thread.sleep(25)
    }
    error("timed out waiting for pid $pid to exit")
}

internal fun invokePing(channel: ManagedChannel, message: String): JsonObject =
    json.parseToJsonElement(
        ClientCalls.blockingUnaryCall(
            channel,
            pingMethod,
            CallOptions.DEFAULT.withDeadlineAfter(2, TimeUnit.SECONDS),
            """{"message":"$message"}""",
        ),
    ).jsonObject

private fun stringMarshaller() = object : MethodDescriptor.Marshaller<String> {
    override fun stream(value: String): InputStream =
        ByteArrayInputStream(value.toByteArray(StandardCharsets.UTF_8))

    override fun parse(stream: InputStream): String =
        stream.readAllBytes().toString(StandardCharsets.UTF_8)
}

private fun shellQuote(value: String): String = "'" + value.replace("'", "'\"'\"'") + "'"
